### DISCUSSION TITLE
Use a buffer for text inputs and textareas instead of binding onValueChanged directly to pushToMetadataValueUpdateQueue

### DISCUSSION THREAD

**Discussion by Audio951** [mProjectsCode/obsidian-meta-bind-plugin#33] [Ideas]
A potential work around for the metadata updates causing rollbacks in state could be to keep all changes in some kind of buffer and wait for inactivity or focusout to apply them. I'm not certain what other fallout this might introduce, but figured I'd bring it up and see what people think. This might also work for toggles ect but idk.

**Comment by mProjectsCode** [comment]
The changes are being buffered in `0.3.2` and since `0.3.2` I haven't had any issues with rollback anymore. 

Should you still experience rollback in `0.3.2` then please let me know how to reproduce it.

**Reply by Audio951** [reply]
Ok, I'll grab latest and test it. Thanks!

**Reply by mProjectsCode** [reply]
I can try to explain how the buffer algorithm works.

There are two buffers, one for `metadata -> input field` and one for `input field -> metadata`. Every fixed interval (sync interval from the settings) the plugin will apply the changes from those two buffers. If there are changes in the `input field -> metadata` buffer, they will be applied and any changes in the `metadata -> input field` buffer will be discarded. Only if there a no changes in the `input field -> metadata` buffer, the `metadata -> input field` buffer will be applied. Meaning that changes in the input field will always take priority over changes in the metadata.

**Reply by Audio951** [reply]
One thing I've noticed so far is that if you make new inputs and never actually interact with them, they are in applyInputFieldValueUpdateQueue every time you interact with any other input. So I had large lists of toggles and text inputs for spells that caused there to be dozens of empty value updates in the queue. Let me know if you need more info there. I'm still investigating what is going on. I could send you my vault if you want to mess with it.

**Reply by Audio951** [reply]
I've run into the metadata being wiped out quite a few times as well. I suspect the metadata update queue system is being overwhelmed somehow.

**Reply by Audio951** [reply]
Also, it seems like there are instances where every input field is updated when modifying one of them, not sure why. If you have discord or something I'd be happy to chat and share there since discussion here can be a little disjointed. Up to you!

**Reply by mProjectsCode** [reply]
Sure message me on discord Lemons#5466

> I've run into the metadata being wiped out quite a few times as well.
I know the issue, but I haven't come up with a good fix for it

**Reply by Audio951** [reply]
You may have to add me, says you're not accepting friend requests.

Audio#1616

**Reply by mProjectsCode** [reply]
Hmmm, server members should be able to send friend requests. Or are you not on the obsidian discord server?
